File: python/src/ecole/setcover.py
import numpy as np
import scipy
import scipy.sparse

from pyscipopt import Model

import ecole.scip


class Setcover:
    def __init__(self, parameter_generator):
        self.parameter_generator = parameter_generator
        self.rng = np.random.RandomState()

    def __iter__(self):
        return self

    def __next__(self):
        param_dict = next(self.parameter_generator)
        model = self.generate_instance(**param_dict)

        return ecole.scip.Model.from_pyscipopt(model)

    def seed(self, seed):
        self.rng.seed(seed)

    def generate_instance(self, nrows=500, ncols=1000, density=0.05, max_coef=100):
        """
		Generate a set cover instance with specified characteristics, and writes
		it to a file in the LP format.
		Algorithm described in:
			E.Balas and A.Ho, Set covering algorithms using cutting planes, heuristics,
			and subgradient optimization: A computational study, Mathematical
			Programming, 12 (1980), 37-60.
		Parameters
		----------
		nrows : int
			Desired number of rows
		ncols : int
			Desired number of columns
		density: float between 0 (excluded) and 1 (included)
			Desired density of the constraint matrix
		max_coef: int
			Maximum objective coefficient (>=1)
		"""
        nnzrs = int(nrows * ncols * density)

        assert nnzrs >= nrows  # at least 1 col per row
        assert nnzrs >= 2 * ncols  # at leats 2 rows per col

        indices = np.empty((nnzrs,), dtype=int)

        # sample column indices
        indices[: 2 * ncols] = np.arange(2 * ncols) % ncols  # force at leats 2 rows per col
        indices[2 * ncols :] = (
            self.rng.choice(ncols * (nrows - 2), size=nnzrs - (2 * ncols), replace=False) % ncols
        )  # remaining column indexes are random

        # count the resulting number of rows, for each column
        _, col_nrows = np.unique(indices, return_counts=True)

        # for each column, sample row indices
        i = 0
        indptr = [0]
        indices[:nrows] = self.rng.permutation(nrows)  # pre-fill to force at least 1 column per row
        for n in col_nrows:

            # column is already filled, nothing to do
            if i + n <= nrows:
                pass

            # column is empty, fill with random rows
            elif i >= nrows:
                indices[i : i + n] = self.rng.choice(nrows, size=n, replace=False)

            # column is partially filled, complete with random rows among remaining ones
            elif i + n > nrows:
                remaining_rows = np.setdiff1d(
                    np.arange(nrows), indices[i:nrows], assume_unique=True
                )
                indices[nrows : i + n] = self.rng.choice(
                    remaining_rows, size=i + n - nrows, replace=False
                )

            i += n
            indptr.append(i)

        # sample objective coefficients
        c = self.rng.randint(max_coef, size=ncols) + 1

        # sparse CSC to sparse CSR matrix
        A = scipy.sparse.csc_matrix(
            (np.ones(len(indices), dtype=int), indices, indptr), shape=(nrows, ncols)
        ).tocsr()
        indices = A.indices
        indptr = A.indptr

        # generate SCIP instance from problem
        model = Model()
        model.setMinimize()

        # add variables
        for j in range(ncols):
            model.addVar(name=f"x{j+1}", vtype="B", obj=c[j])

        # add constraints
        model_vars = model.getVars()
        for i in range(nrows):
            cons_lhs = 0
            consvars = [model_vars[j] for j in indices[indptr[i] : indptr[i + 1]]]
            for var in consvars:
                cons_lhs += var
            model.addCons(cons_lhs >= 1)

        return model

File: libecole/src/instance/set-cover.cpp
#include <fmt/format.h>
#include <map>

#include <xtensor/xrandom.hpp>
#include <xtensor/xsort.hpp>
#include <xtensor/xtensor.hpp>
#include <xtensor/xview.hpp>

#include "ecole/instance/set-cover.hpp"
#include "ecole/scip/cons.hpp"
#include "ecole/scip/model.hpp"
#include "ecole/scip/utils.hpp"
#include "ecole/scip/var.hpp"

namespace ecole::instance {

/*************************************
 *  SetCoverGenerator methods  *
 *************************************/

SetCoverGenerator::SetCoverGenerator(Parameters parameters_, RandomEngine random_engine_) :
	random_engine{random_engine_}, parameters{parameters_} {}
SetCoverGenerator::SetCoverGenerator(Parameters parameters_) :
	SetCoverGenerator{parameters_, ecole::spawn_random_engine()} {}
SetCoverGenerator::SetCoverGenerator() : SetCoverGenerator(Parameters{}) {}

scip::Model SetCoverGenerator::next() {
	return generate_instance(parameters, random_engine);
}

void SetCoverGenerator::seed(Seed seed) {
	random_engine.seed(seed);
}

namespace {

using std::size_t;
using xvector = xt::xtensor<size_t, 1>;

/** Sets a slice of a 1-D xtensor.
 *
 * Sets the the slice from start_index to end_index of vector a to
 * vector b.  Vector b should be of size (end_index - start_index).
 */
auto set_slice(xvector& a, xvector& b, size_t start_index, size_t end_index) {
	auto slice = xt::view(a, xt::range(start_index, end_index));
	slice = b;
}

/** Gets a slice of a 1-D xtensor.
 *
 * Gets the the slice from start_index to end_index of vector a.
 */
auto get_slice(xvector& a, size_t start_index, size_t end_index) -> xvector {
	auto slice = xt::view(a, xt::range(start_index, end_index));
	return slice;
}

/** Gets the counts of each element in a 1-D xtensor.
 *
 * This method is specifically for the case where
 * each each element is in the range [0, n_cols].
 */
auto get_counts(xvector& indices, size_t n_cols) -> xvector {
	xvector counts({n_cols}, 0);
	for (size_t i = 0; i < indices.size(); ++i) {
		++counts(indices[i]);
	}
	return counts;
}

/** Samples values in a range and returns them as a 1-D xtensor.
 *
 * Samples num_samples values in the range from start_index to
 * end_index.
 */
auto get_choice_in_range(size_t start_index, size_t end_index, size_t num_samples, RandomEngine& random_engine)
	-> xvector {
	xvector choices = xt::arange<size_t>(start_index, end_index, 1);
	xvector samples = xt::random::choice(choices, num_samples, false, random_engine);
	return samples;
}

/** Adds a varaible to the SCIP Model.
 *
 * Adds a single binary variable with a specified cost.
 */
auto add_var(SCIP* scip, size_t idx, SCIP_Real cost) -> SCIP_VAR* {
	auto const name = fmt::format("x_{}", idx);
	auto unique_var = scip::create_var_basic(scip, name.c_str(), 0., 1., cost, SCIP_VARTYPE_BINARY);
	auto* var_ptr = unique_var.get();
	scip::call(SCIPaddVar, scip, var_ptr);
	return var_ptr;
}

/** Adds all varaibles to the SCIP Model.
 *
 * A variable is added for each element (or column) in the
 * set cover problem.
 */
auto add_vars(SCIP* scip, xt::xtensor<SCIP_Real, 1> const& c) -> xt::xtensor<SCIP_VAR*, 1> {
	auto vars = xt::xtensor<SCIP_VAR*, 1>{{c.shape()}};
	for (size_t i = 0; i < c.size(); ++i) {
		vars(i) = add_var(scip, i, c(i));
	}
	return vars;
}

/** Adds set covering constaints.
 *
 * For each set, at least one element is required to the solution.
 */
auto add_constaints(SCIP* scip, xt::xtensor<SCIP_VAR*, 1> vars, xvector& indices, xvector& indptr, size_t n_rows) {

	for (size_t i = 0; i < n_rows; ++i) {

		// get the set of variables to be added in the constraint.
		auto cons_vars = xt::xtensor<SCIP_VAR*, 1>{{indptr(i + 1) - indptr(i)}};
		for (size_t j = indptr(i); j < indptr(i + 1); ++j) {
			cons_vars(j - indptr(i)) = vars(indices(j));
		}

		// add constraint to SCIP Model.
		auto name = fmt::format("c_{}", i);
		auto const inf = SCIPinfinity(scip);
		auto coefs = xt::xtensor<SCIP_Real, 1>(cons_vars.shape(), 1.);
		auto cons =
			scip::create_cons_basic_linear(scip, name.c_str(), cons_vars.size(), &cons_vars(0), coefs.data(), 1.0, inf);
		scip::call(SCIPaddCons, scip, cons.get());
	}
}

/** Convert CSC sparse indicies and index pointers to CSR.
 *
 * This implementation only converts the indices and points,
 * and does not consider the values as they are all 1.  A
 * tuple of the indices and index pointers in CSR format are
 * returned.
 */
auto convert_csc_to_csr(xvector& indices, xvector& indptr, size_t n_rows, size_t n_cols) {

	xvector indptr_csr({n_rows + 1}, 0);
	xvector indices_csr({indices.size()}, 0);

	for (size_t j = 0; j < indices.size(); ++j) {
		indptr_csr(indices(j) + 1) += 1;
	}
	indptr_csr = xt::cumsum(indptr_csr);

	for (size_t col = 0; col < n_cols; ++col) {
		for (size_t jj = indptr(col); jj < indptr(col + 1); ++jj) {
			size_t row = indices(jj);
			indices_csr(indptr_csr(row)) = col;
			indptr_csr(row) = indptr_csr(row) + 1;
		}
	}

	size_t last = 0;
	for (size_t row = 0; row <= n_rows; ++row) {
		std::swap(last, indptr_csr[row]);
	}

	return std::make_tuple(indptr_csr, indices_csr);
}

}  // namespace

/******************************************
 *  SetCoverGenerator::generate_instance  *
 ******************************************/

scip::Model SetCoverGenerator::generate_instance(Parameters parameters, RandomEngine& random_engine) {

	auto const n_rows = parameters.n_rows;
	auto const n_cols = parameters.n_cols;
	auto const density = parameters.density;
	auto const max_coef = static_cast<size_t>(parameters.max_coef);

	auto const nnzrs = static_cast<size_t>(static_cast<double>(n_rows * n_cols) * density);

	xvector indices({nnzrs}, 0);

	// force at least 2 rows per col
	xvector first_indices = xt::arange<size_t>(0, 2 * n_cols, 1) % n_cols;
	set_slice(indices, first_indices, 0, 2 * n_cols);

	// assign remaining column indexes at random
	xvector samples = get_choice_in_range(0, n_cols * (n_rows - 2), nnzrs - (2 * n_cols), random_engine) % n_cols;
	set_slice(indices, samples, 2 * n_cols, nnzrs);

	// get counts of unique elements
	auto col_n_rows = get_counts(indices, n_cols);

	// ensure at least 1 column per row
	auto perm = xt::random::permutation<size_t>(n_rows, random_engine);
	set_slice(indices, perm, 0, n_rows);

	size_t i = 0;
	xvector indptr({n_cols + 1}, 0);
	size_t indptr_idx = 1;

	for (size_t idx = 0; idx < col_n_rows.size(); ++idx) {
		size_t n = col_n_rows(idx);

		if (i + n <= n_rows) {
		} else if (i >= n_rows) {
			xvector sampled_rows = get_choice_in_range(0, n_rows, n, random_engine);
			set_slice(indices, sampled_rows, i, i + n);

		} else if (i + n > n_rows) {
			auto remaining_rows = xt::setdiff1d(xt::arange<size_t>(0, n_rows, 1), get_slice(indices, i, n_rows));
			xvector choices = xt::random::choice(remaining_rows, i + n - n_rows, false, random_engine);
			set_slice(indices, choices, n_rows, i + n);
		}

		i += n;
		indptr(indptr_idx) = i;
		++indptr_idx;
	}

	// convert csc indices/ptrs to csr
	auto [indptr_csr, indices_csr] = convert_csc_to_csr(indices, indptr, n_rows, n_cols);

	// sample coefficients
	xt::xtensor<SCIP_Real, 1> c = xt::random::randint<size_t>({n_cols}, 0, max_coef, random_engine);

	// create scip model
	auto model = scip::Model::prob_basic();
	model.set_name(fmt::format("SetCover-{}-{}", parameters.n_rows, parameters.n_cols));
	auto* const scip = model.get_scip_ptr();
	scip::call(SCIPsetObjsense, scip, SCIP_OBJSENSE_MINIMIZE);

	// add variables and constraints
	auto const vars = add_vars(scip, c);
	add_constaints(scip, vars, indices_csr, indptr_csr, n_rows);

	return model;

}  // generate_instance

}  // namespace ecole::instance
